ops/model.py

```diff
@@ -176,20 +245,26 @@ def get_binding(self, binding_key: typing.Union[str, 'Relation']) -> 'Binding':
         return self._bindings.get(binding_key)
 
 
-class _ModelCache:
+_T = TypeVar('_T', bound='UnitOrApplication')
+
 
+class _ModelCache:
     def __init__(self, meta: 'ops.charm.CharmMeta', backend: '_ModelBackend'):
+        if typing.TYPE_CHECKING:
+            _weakcachetype = weakref.WeakValueDictionary[
+                Tuple[Any, ...], Optional[UnitOrApplication]]
+
         self._meta = meta
         self._backend = backend
-        self._weakrefs = weakref.WeakValueDictionary()
+        self._weakrefs = weakref.WeakValueDictionary()  # type: _weakcachetype
 
-    def get(self, entity_type, *args):
+    def get(self, entity_type: Type['_T'], *args: Any) -> '_T':
```

Comment:
This and the above: I hesitated to refactor because I thought this class may have some (intended) usage I was unaware of. ATM it's too generic, I agree.